apis/common/v1/resource.go

```diff
@@ -77,10 +77,47 @@ type SecretKeySelector struct {
 	Key string `json:"key"`
 }
 
+// A ReferenceResolutionPolicy is a value for resolution policy for reference.
+type ReferenceResolutionPolicy string
+
+const (
+	// ReferencePolicyOptional is a resolution option.
+	// When the ReferenceResolutionPolicy is set to ReferencePolicyOptional the
+	// execution could continue even if the reference cannot be resolved.
+	ReferencePolicyOptional ReferenceResolutionPolicy = "Optional"
+
+	// ReferencePolicyAlways is a resolution option.
+	// When the ReferenceResolutionPolicy is set to ReferencePolicyAlways the
+	// reference will be tried to resolve for every reconcile loop.
+	ReferencePolicyAlways ReferenceResolutionPolicy = "Always"
+)
+
 // A Reference to a named object.
 type Reference struct {
```

Comment:
The policies here affect `ref -> value` transition but there is also `selector -> ref` where similar policies could apply. But as far as I can tell, both issues are about `ref -> value`, so having the policies only here could make sense for now until users see a need to customize `selector -> ref` as well.

Reply:
Doesn't for example https://github.com/crossplane/crossplane-runtime/issues/250#issuecomment-912800805 pertain to selectors? I could imagine the race @chlunde describes being a problem here, i.e.:

1. Foo A is created.
2. fooSelector selects Foo A and creates a reference.
3. Foo B is created
4. fooSelector would have selected Foo B, but doesn't because it already successfully selected its references.

Reply:
Thank you for bringing up this, it was a point that I thought while working on these issues. Implemented supports/API for the `Reference` field can also be provided for `Selector` field, as you said. However, at this point, I think a new discussion can start.

Let me share my thoughts via an example:

```yaml
..
forProvider:
  vpcIdSelector:
    matchLabels:
      key: value
..
```

As you know, while we using the Selector field for resolution, we are filling the Reference field. So after we selected and resolved the reference by using the Selector, new state of the resource will be like that:

```yaml
..
forProvider:
  vpcId: vpc-1111111111
  vpcIdRef:
    name: sample-vpc
  vpcIdSelector:
    matchLabels:
      key: value
..
```

Now, when we review the [function] (this is Resolve function), we are observing that, firstly the Reference field is checked and then if it is not empty, the resolution process is done by using the Reference field and we return early. In other words, if the Reference field is set, we do not process the Selector field. Therefore, after this point, even if I update the Selector field, it will have no effect. Because the Reference field is full and only that field will be processed in the next loops.

I think this is an independent point from the resolution/resolve policies. Because, at this point, issues such as which resolution method is more dominant and which one to choose for situations where both are set will come to the fore. Let me share an example to clarify my point. Imagine that a user tries to provision an external resource by using the following manifest:

```yaml
..
forProvider:
  vpcIdRef:
    name: sample-vpc
  vpcIdSelector:
    matchLabels:
      key: value
..
```

For a resource that is being create by this manifest, we can say that the reference in the Reference field is dominant. So, the Selector field will not be used for resolution process. In fact, this is the same situation we encountered after selecting references and filling Reference field via Selector. And from this point on, the Selector field will not be considered again. Therefore, changes on this field, such as updating list or changing the policy, will not be affective.

As a result, I think providing support of new Policy API for Selector field makes sense. However, if I do not miss any point, it seems that we will have to do this with a solution of the point I tried to explain above.

[function]: https://github.com/crossplane/crossplane-runtime/blob/0e8935d6d9a942825a29ab3ab34f057cfb1dc050/pkg/reference/reference.go#L210

Reply:
@negz Yes, you're right. We need to address the selector as well.

> As a result, I think providing support of new Policy API for Selector field makes sense. However, if I do not miss any point, it seems that we will have to do this with a solution of the point I tried to explain above.

@sergenyalcin If I understood you correctly, you're saying that the fact that we don't take any action after the ref is found will be changed. I think that's fine since it's configured by the user.